File: sdk/typescript-mastra/src/workspace.ts
import { randomUUID } from "node:crypto";
import type { ChannelOptions } from "@grpc/grpc-js";
import { Client, type ClientOptions } from "@neuronedge/enclave";

import { mapServiceError } from "./errors.js";
import { createEnclaveTools } from "./tools.js";

/** Options for {@link EnclaveWorkspace}. */
export type EnclaveWorkspaceOptions = {
  /** gRPC target, e.g. `"127.0.0.1:50051"`. */
  target: string;
  /** Default: `agent-${crypto.randomUUID().replace(/-/g, "")}`. */
  workspaceId?: string;
  /** Default 2. */
  vcpuCount?: number;
  /** Default 1024. */
  memSizeMib?: number;
  /** Default: `NE_KERNEL_IMAGE_PATH`. */
  kernelImagePath?: string;
  /** Default: `NE_ROOTFS_IMAGE_PATH`. */
  rootfsImagePath?: string;
  /** Default: `NE_VSOCK_CID_BASE` (no silent fallback — must be unique per concurrent workspace on a host). */
  guestVsockCid?: number;
  /** Default false — agents need to write. (The SDK server default is true.) */
  rootfsReadOnly?: boolean;
  kernelBootArgs?: string;
  /** Forwarded verbatim to the underlying base SDK channel. */
  channelOptions?: ChannelOptions;
  /** Default 2000. */
  destroyGracePeriodMs?: number;
  /** @internal Test injection — substitutes the base `Client`. */
  _clientFactory?: () => Client;
};

/** Emit a teardown warning (a destroy failure must not crash a successful
 *  block, and must not mask the caller's exception). Uses `console.warn`
 *  deliberately — STANDARDS' "no console.log" targets debug logging; a genuine
 *  leak/teardown warning is what `console.warn` is for. Mirrors 7.5's
 *  `logging.getLogger("ne_langchain").warning(...)`. */
function warnTeardownFailure(workspaceId: string, err: unknown): void {
  const detail = err instanceof Error ? err.message : String(err);
  console.warn(`[neuronedge-enclave-mastra] destroyWorkspace failed for ${workspaceId}: ${detail}`);
}

/** Owns one Firecracker microVM workspace across `start()` / `stop()`. */
export class EnclaveWorkspace {
  private readonly _opts: EnclaveWorkspaceOptions;
  private readonly _workspaceId: string;
  private _client: Client | null = null;
  private _started = false;
  private _stopped = false;

  constructor(options: EnclaveWorkspaceOptions) {
    this._opts = options;
    this._workspaceId = options.workspaceId ?? `agent-${randomUUID().replace(/-/g, "")}`;
  }

  /** Workspace id (the constructor-generated default is available before start). */
  get workspaceId(): string {
    return this._workspaceId;
  }

  /** The base SDK client (passthrough for power users — snapshot/attest/etc.). */
  get client(): Client {
    if (this._client === null) {
      throw new Error("EnclaveWorkspace not started; call start() (or use withWorkspace).");
    }
    return this._client;
  }

  /** The three tools bound to this workspace. Throws before `start()`. */
  get tools() {
    if (!this._started || this._client === null) {
      throw new Error("EnclaveWorkspace not started; call start() (or use withWorkspace).");
    }
    return createEnclaveTools(this._client, this.workspaceId);
  }

  /** Resolve env defaults, validate, construct a base `Client`, and
   *  `createWorkspace`. Throws before any RPC if required inputs are missing. */
  async start(): Promise<this> {
    if (this._started) return this;

    const workspaceId = this._workspaceId;

    const kernelImagePath = this._opts.kernelImagePath ?? process.env.NE_KERNEL_IMAGE_PATH;
    const rootfsImagePath = this._opts.rootfsImagePath ?? process.env.NE_ROOTFS_IMAGE_PATH;
    const cidEnv = process.env.NE_VSOCK_CID_BASE;
    const guestVsockCid =
      this._opts.guestVsockCid ?? (cidEnv !== undefined ? Number(cidEnv) : undefined);

    const missing: string[] = [];
    if (!kernelImagePath) missing.push("kernelImagePath");
    if (!rootfsImagePath) missing.push("rootfsImagePath");
    if (guestVsockCid === undefined || Number.isNaN(guestVsockCid)) missing.push("guestVsockCid");
    if (missing.length > 0) {
      throw new Error(
        `EnclaveWorkspace missing required inputs (pass as options or set NE_* env): ${missing.join(", ")}`,
      );
    }

    const clientOptions: ClientOptions = {
      target: this._opts.target,
      ...(this._opts.channelOptions !== undefined
        ? { channelOptions: this._opts.channelOptions }
        : {}),
    };
    let client: Client;
    if (this._opts._clientFactory) {
      client = this._opts._clientFactory();
    } else {
      // Production-only path: a real Client needs a live gRPC server. The
      // _clientFactory seam is how unit tests substitute it; this arm is
      // exercised by examples/quickstart.ts + the base SDK's Client tests.
      /* v8 ignore next */
      client = new Client(clientOptions);
    }

    try {
      await client.createWorkspace({
        workspaceId,
        kernelImagePath: kernelImagePath!, // presence validated in `missing` above
        rootfsImagePath: rootfsImagePath!, // presence validated in `missing` above
        vcpuCount: this._opts.vcpuCount ?? 2,
        memSizeMib: this._opts.memSizeMib ?? 1024,
        guestVsockCid: guestVsockCid!, // presence + NaN validated in `missing` above
        ...(this._opts.rootfsReadOnly !== undefined
          ? { rootfsReadOnly: this._opts.rootfsReadOnly }
          : { rootfsReadOnly: false }),
        ...(this._opts.kernelBootArgs !== undefined
          ? { kernelBootArgs: this._opts.kernelBootArgs }
          : {}),
      });
    } catch (err) {
      client.close();
      mapServiceError(err); // always throws — surfaces "enclave RPC <STATUS>: <details>"
    }

    this._client = client;
    this._started = true;
    return this;
  }

  /** Destroy the workspace + close the client. Idempotent. Best-effort: the
   *  swallow/preserve contract belongs to {@link withWorkspace}. */
  async stop(): Promise<void> {
    if (this._stopped || !this._started || this._client === null) return;
    // Set before the RPC + null the client in `finally`: if destroyWorkspace
    // rejects, stop() is unrecoverable — the caller can neither retry stop()
    // nor reach ws.client to destroy manually. Intentional for a dev-pilot:
    // destroy is a terminal best-effort, and operator reaping (the
    // supervisor's `grace_period_ms`) is the recovery path.
    this._stopped = true;
    try {
      await this._client.destroyWorkspace({
        workspaceId: this.workspaceId,
        gracePeriodMs: this._opts.destroyGracePeriodMs ?? 2000,
      });
    } finally {
      this._client.close();
      this._client = null;
    }
  }

  /** `await using` forward-compat (Node 22+). Delegates to `stop()`. */
  async [Symbol.asyncDispose](): Promise<void> {
    await this.stop();
  }
}

/** Primary API: the closest JS analogue of Python's `with`. Boots a workspace,
 *  runs `fn` with its tools, and guarantees teardown on both paths. A destroy
 *  failure on the success path is logged + swallowed; on the exception path the
 *  **original** caller exception always wins (never masked by a destroy error). */
export async function withWorkspace<T>(
  options: EnclaveWorkspaceOptions,
  fn: (ws: EnclaveWorkspace) => Promise<T>,
): Promise<T> {
  const ws = new EnclaveWorkspace(options);
  await ws.start(); // if start rejects, fn is never invoked and the rejection propagates

  let result: T | undefined;
  let callerError: unknown;
  let threw = false;
  try {
    result = await fn(ws);
  } catch (e) {
    callerError = e;
    threw = true;
  }

  try {
    await ws.stop();
  } catch (destroyErr) {
    warnTeardownFailure(ws.workspaceId, destroyErr);
  }

  if (threw) throw callerError;
  return result as T;
}

File: sdk/python-langchain/src/ne_langchain/workspace.py
"""Managed-workspace lifecycle for the LangChain adapter."""

from __future__ import annotations

import logging
import os
import uuid
from collections.abc import Iterable
from types import TracebackType

import ne

from ne_langchain.tools import EnclaveToolkit

_log = logging.getLogger("ne_langchain")


class EnclaveWorkspace:
    """Owns one Firecracker microVM workspace for the lifetime of a ``with`` block.

    On enter: resolves kernel/rootfs digests and vsock-cid from kwargs or ``NE_*`` env
    vars (raising ``ValueError`` if any are missing), opens a base-SDK
    ``ne.Client`` (insecure channel in this phase), and calls
    ``create_workspace``. On exit: always calls ``destroy_workspace``;
    teardown failures on the success path are logged and swallowed, on the
    exception path the original exception wins.

    Example::

        with EnclaveWorkspace(target="127.0.0.1:50051") as ws:
            tools = ws.tools.get_tools()
    """

    def __init__(
        self,
        target: str,
        *,
        workspace_id: str | None = None,
        vcpu_count: int = 2,
        mem_size_mib: int = 1024,
        kernel_sha256: str | None = None,
        rootfs_sha256: str | None = None,
        guest_vsock_cid: int | None = None,
        rootfs_read_only: bool = False,
        kernel_boot_args: str | None = None,
        channel_options: Iterable[tuple[str, object]] | None = None,
        destroy_grace_period_ms: int = 2_000,
    ) -> None:
        self._target = target
        self._workspace_id = workspace_id or f"agent-{uuid.uuid4().hex}"
        self._vcpu_count = vcpu_count
        self._mem_size_mib = mem_size_mib
        self._kernel_sha256 = kernel_sha256
        self._rootfs_sha256 = rootfs_sha256
        self._guest_vsock_cid = guest_vsock_cid
        self._rootfs_read_only = rootfs_read_only
        self._kernel_boot_args = kernel_boot_args
        self._channel_options = channel_options
        self._destroy_grace_period_ms = destroy_grace_period_ms
        self._client: ne.Client | None = None

    @property
    def workspace_id(self) -> str:
        return self._workspace_id

    @property
    def client(self) -> ne.Client:
        if self._client is None:
            raise RuntimeError(
                "EnclaveWorkspace not entered; use 'with EnclaveWorkspace(...) as ws:'"
            )
        return self._client

    @property
    def tools(self) -> EnclaveToolkit:
        return EnclaveToolkit(client=self.client, workspace_id=self._workspace_id)

    def __enter__(self) -> EnclaveWorkspace:
        kernel = self._kernel_sha256 or os.environ.get("NE_KERNEL_SHA256")
        rootfs = self._rootfs_sha256 or os.environ.get("NE_ROOTFS_SHA256")
        cid_env = os.environ.get("NE_VSOCK_CID_BASE")
        cid = (
            self._guest_vsock_cid
            if self._guest_vsock_cid is not None
            else (int(cid_env) if cid_env else None)
        )

        missing = [
            name
            for name, value in (
                ("kernel_sha256", kernel),
                ("rootfs_sha256", rootfs),
                ("guest_vsock_cid", cid),
            )
            if not value
        ]
        if missing:
            raise ValueError(
                "EnclaveWorkspace missing required inputs (pass as kwargs or set NE_* env): "
                + ", ".join(missing)
            )

        client = ne.Client(self._target, channel_options=list(self._channel_options or ()))
        try:
            client.create_workspace(
                workspace_id=self._workspace_id,
                kernel_sha256=kernel,
                rootfs_sha256=rootfs,
                vcpu_count=self._vcpu_count,
                mem_size_mib=self._mem_size_mib,
                guest_vsock_cid=int(cid),  # type: ignore[arg-type]
                rootfs_read_only=self._rootfs_read_only,
                kernel_boot_args=self._kernel_boot_args,
            )
        except BaseException:
            client.close()
            raise
        self._client = client
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        client = self._client
        if client is None:
            return None
        try:
            client.destroy_workspace(
                workspace_id=self._workspace_id,
                grace_period_ms=self._destroy_grace_period_ms,
            )
        except BaseException as destroy_err:
            # Never re-raise: on the success path we swallow after logging;
            # on the exception path swallowing lets the caller's original
            # exception propagate unmasked (a re-raise here would replace it).
            _log.warning("destroy_workspace failed for %s: %s", self._workspace_id, destroy_err)
        finally:
            client.close()
            self._client = None
        return None
